docs: add ivy opt-in docs (#28569)
angular/angular: aio/content/cli/ivy.md

> # Using Ivy with Angular CLI

Author: What is the best place to add this page in?

Reviewer: not sure. @jenniferfell can you please propose some place?

Reviewer 2: This doc should be placed in the `aio/content/guide` directory. I think it would be good also to rename the doc to something like `ivy-preview` if this is going to be temporary.

Author: Renamed and added under the `Techniques` section of navigation as `Ivy Preview`. @StephenFluin  @jenniferfell can you confirm this location is the appropriate one?

Starting with Angular version 8, you can experimentally try out the new [Ivy Compiler](link to some ivy announcement here).

## Starting a new project using Ivy

The easiest way to start a new project with Ivy is to use the `--experimental-ivy` flag on `ng new`:

```
ng new shiny-ivy-app --experimental-ivy
```

Everything will be configured for you.

{@a updating}
## Updating an existing project to use Ivy

By configuring a few key files, you can also update your existing project to use Ivy.

- add the `enableIvy` under `angularCompilerOptions` in `src/tsconfig.app.json`:
```json
{

>   "compilerOptions": { ... },

Author: Is using an ellipsis an ok way of representing hidden content for partial options?

Reviewer: yes

  "angularCompilerOptions": {

>     "enableIvy": true,

Reviewer 3: Is true the recommended value here?  The value currently is `ngtsc` with the `experimentalIvy` flag

Reviewer 4: Yeah, I don't think `true` works right now. It should be `'ngtsc'`.

Author: The intend is that it works with `true/false`. I have another task to make those changes.

Reviewer 4: 👍 Let's just make sure we don't merge this doc until that other task has been completed 😉

Reviewer: true/false is what we want, as @filipesilva mentioned we are landing a change to use simple boolean values for the flag.

Author: https://github.com/angular/angular/pull/28616 is now merged.

  }
}
```
- set the default build options for your project to always use AOT compilation if it isn't already:
```json
{
  "projects": {
    "my-existing-project": {
      "architect": {
        "build": {
          "options": {
            ...
            "aot": true,

> ```

Author: Is it better to close all the brackets, or to not use the `json` markdown hint?

Reviewer: close all brackets plz

Author: Done

- add a `postinstall` script in `package.json` to run the [Angular Compatibility Compiler](#ngcc):
```json
{
  "scripts": {
    ...
    "postinstall": "ivy-ngcc"
  }
}
```
- reinstall your node modules to run the newly added script.

## Switching back to the current compiler

To stop using the Ivy compiler you need to undo the steps taken when [updating to use Ivy](#updating).
- set `enableIvy` to false in `src/tsconfig`, or remove it completely.
- remove `"aot": true` from your default build options if you didn't have it there before.
- remove the `postinstall` script.
- delete and reinstall your node modules.

{@a ngcc}
## The Angular Compatibility Compiler

Your third party libraries also need to be compiled with Ivy for you to use them in your Ivy application.
That's where the Angular Compatibility compiler comes into play.

We set it as a postinstall script in your `package.json` so that it always runs when you install

> new modules.

Reviewer: Please add: Using postinstall hook to run ngcc is just a temporary integration. We expect ngcc to be seamlessly integrated into the Angular CLI build pipeline in the future before the full Ivy rollout. Once that's implemented ngcc will and not be visible to developers.

Author: Done.

        }
      }
    }
  }
}
```
- add a `postinstall` script in `package.json` to run the [Angular Compatibility Compiler](#ngcc):
```json
{
  "scripts": {
    ...
    "postinstall": "ivy-ngcc"
  }
}
```
- reinstall your node modules to run the newly added script.

## Switching back to the current compiler

To stop using the Ivy compiler you need to undo the steps taken when [updating to use Ivy](#updating).
- set `enableIvy` to false in `src/tsconfig`, or remove it completely.
- remove `"aot": true` from your default build options if you didn't have it there before.
- remove the `postinstall` script.
- delete and reinstall your node modules.

{@a ngcc}
## The Angular Compatibility Compiler

Your third party libraries also need to be compiled with Ivy for you to use them in your Ivy application.
That's where the Angular Compatibility (`ngcc`) compiler comes into play.

We set it as a postinstall script in your `package.json` so that it always runs when you install
new modules.

Using a postinstall hook to run `ngcc` is just a temporary integration. 
We expect `ngcc` to be seamlessly integrated into the Angular CLI build pipeline in the future before the full Ivy rollout. 

> Once that's implemented `ngcc` will and not be visible to developers.

Reviewer 5: @filipesilva, this is really good!

Should we mentioned the limitation that if one has multiple projects within the workspace, all of them will need to be build using Ivy? As users will have cryptic error messages on projects that rely on `ngc` but `ngcc` was run.

Author: Added a note about that:

> Until that happens, opting into Ivy means that all projects in a single CLI workspace will be compiled with Ivy.

# Using Ivy with Angular CLI

Starting with Angular version 8, you can experimentally try out the new [Ivy Compiler](link to some ivy announcement here).

## Starting a new project using Ivy

The easiest way to start a new project with Ivy is to use the `--experimental-ivy` flag on `ng new`:

```
ng new shiny-ivy-app --experimental-ivy
```

Everything will be configured for you.

{@a updating}
## Updating an existing project to use Ivy

By configuring a few key files, you can also update your existing project to use Ivy.

- add the `enableIvy` under `angularCompilerOptions` in `src/tsconfig.app.json`:
```json
{
  "compilerOptions": { ... },
  "angularCompilerOptions": {

>     "enableIvy": true,

Reviewer 4: Isn't `"allowEmptyCodegenFiles": true` required any more?

Author: Yes, it is still required right now. But https://github.com/angular/angular/pull/28685 (or a similar approach) will make it not needed.

Author: Added it here for now, to remove later when the other PR is in.

# Using Ivy with Angular CLI

Starting with Angular version 8, you can experimentally try out the new [Ivy Compiler](link to some ivy announcement here).

## Starting a new project using Ivy

> The easiest way to start a new project with Ivy is to use the `--experimental-ivy` flag on `ng new`:

Reviewer 2: ```suggestion
To start a new project with Ivy enabled, use the `--experimental-ivy` flag with the [`ng new`](cli/new) command:
```

> ```

Reviewer 2: ```suggestion
```sh
```

ng new shiny-ivy-app --experimental-ivy
```

> Everything will be configured for you.

Reviewer 2: I think this should list the steps that will be done similar to the list of items that you have to do to undo the changes.

Author: Done.

{@a updating}
## Updating an existing project to use Ivy

> By configuring a few key files, you can also update your existing project to use Ivy.

Reviewer 2: ```suggestion
By configuring a few key files, you can also update your existing project(s) to use Ivy.
```

> - add the `enableIvy` under `angularCompilerOptions` in `src/tsconfig.app.json`:

Reviewer 2: ```suggestion
- Add the `enableIvy` option in the `angularCompilerOptions` in your project's `src/tsconfig.app.json`:
```

```json
{
  "compilerOptions": { ... },
  "angularCompilerOptions": {
    "enableIvy": true,
    "allowEmptyCodegenFiles": true,
  }
}
```

> - set the default build options for your project to always use AOT compilation if it isn't already:

Reviewer 2: ```suggestion
- Set the default build options for your project to always use AOT compilation if it isn't already:
```

```json
{
  "projects": {
    "my-existing-project": {
      "architect": {
        "build": {
          "options": {
            ...
            "aot": true,
          }
        }
      }
    }
  }
}
```

> - add a `postinstall` script in `package.json` to run the [Angular Compatibility Compiler](#ngcc):

Reviewer 2: ```suggestion
- Add a `postinstall` script to the workspace `package.json` file to run the [Angular Compatibility Compiler](#ngcc):
```

```json
{
  "scripts": {
    ...
    "postinstall": "ivy-ngcc"
  }
}
```

> - reinstall your node modules to run the newly added script.

Reviewer 2: ```suggestion
- Reinstall your `package.json` dependencies to run the newly added script.
```

## Switching back to the current compiler

To stop using the Ivy compiler you need to undo the steps taken when [updating to use Ivy](#updating).

> - set `enableIvy` to false in `src/tsconfig`, or remove it completely.

Reviewer 2: Capitalize each bullet

Author: Done

- remove `"aot": true` from your default build options if you didn't have it there before.
- remove the `postinstall` script.
- delete and reinstall your node modules.

{@a ngcc}
## The Angular Compatibility Compiler

Your third party libraries also need to be compiled with Ivy for you to use them in your Ivy application.
That's where the Angular Compatibility (`ngcc`) compiler comes into play.

We set it as a postinstall script in your `package.json` so that it always runs when you install

> new modules.

Reviewer 2: ```suggestion
install your `package.json` dependencies.
```
